src/lib.rs

```diff
@@ -14,8 +15,77 @@ pub fn eval(
         println!("{}", e.pp(&lexer, &calc_y::token_epp));
     }
     match res {
-        Some(Ok(r)) => Ok(format!("Result: {:?}", r)),
-        // TODO: prvide more infomraiton about the exact error.
+        Some(x) => {
+            return Ok(format!("Result: {}", Eval::new(input).eval(&x)));
+        }
         _ => Err(String::from("Unable to evaluate expression.")),
     }
 }
+
+struct Eval<'a> {
```

Comment:
For the next PR or two I'd like to see a nice split between "compile an AST to `Vec<Opcode>` and then have an evaluator which takes `Vec<Opcode>` and executes the program". [@vext01 does that make sense?]

Reply:
Yes, but I think Pavel was aiming for an AST interpreter.

I have no preference beyond being more familiar with bytecode interpreters...

Reply:
Added "split" to TODO tasks in README https://github.com/softdevteam/pavel.yaiwr/pull/7/commits/bde317ac0f1e3c2ca1ab6cba5d7adfd554d62146

Reply:
This is a *parse tree* interpreter -- actually it's a parse tree evaluator! What we want is an AST compiler.

Quick terminology: "parse tree" includes all the "useless" syntactic information that humans like/need but doesn't affect compilation; "ASTs" strip out that useless syntactic stuff; "evaluator" evaluates something (parse tree, AST, or opcodes) directly; a "compiler" converts one thing into another.

We don't have to do that stuff in this PR but we should move towards a scenario where a) there's a clear split between compiler and interpreter b) we use parse actions (not `GenericParseTree`).

Reply:
I agree.

We spoke about this yesterday on mattermost. I think he is getting away with it because everything is constant, and can be folded at parse time. I don't see how you could implement variables via this approach though.

Reply:
I don't see anything being folded at parse time?

Reply:
Ah, so is code like this happening *after* parsing?

https://github.com/softdevteam/pavel.yaiwr/pull/7/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R40

TBH I have no idea (or desire to know) how grmtools works, so I'm totally winging it.

Reply:
Yes, that's why this is a parse tree evaluator. grmtools produces a parse tree, then this code evaluates using that parse tree.